### PR TITLE
Fix peer authentiation selection bug due to drop in singleton restriction

### PR DESCRIPTION
**Pull request by diemtvu**
As we drop the restriction for the namespace/mesh policy name in https://github.com/istio/istio/pull/21136, the guarantee that there is at most CR that have no selector per namespace is no longer hold. As a result, we need to check all policy and keep the oldest to be consistent.

### CODE REVIEW

**Comment by incfly** [comment]
>  As a result, we need to check all policy and keep the oldest to be consistent.

This effectively is still keeping singleton for different Istio revisions, no? Does it help with the multi control plane thing? @howardjohn

**Comment by howardjohn** [comment]
> This effectively is still keeping singleton for different Istio revisions, no? Does it help with the multi control plane thing? @howardjohn

Its fine, at the very low level of the CRD reading code will filter out config that does not apply to the revision. Each control plane will have a single config, but that config may be different

**Comment by diemtvu** [comment]
> > As a result, we need to check all policy and keep the oldest to be consistent.
> 
> This effectively is still keeping singleton for different Istio revisions, no? Does it help with the multi control plane thing? @howardjohn
> 
> which as I understand is the original motivation to drop the restriction.

The name restriction is removed so that user can have different mesh/namespace wide policy per control plane (istiod). AFAIK, users need to label the config for control plane `red` with `red` label (or smt equivalent), and `red` control plane will see only that config, not the ones for `blue` etc.

However, as the name can be anything, it's possible to have misconfiguration when there are more than one namespace/mesh policy per control plane. Thus we need to treat it in the same way as other: the oldest one win.

I also add another change to reject newer namespace/mesh policy during push context initialization. That makes the change in `policy_applier` redundant, though I think it's still good to have both, just in case.

**Comment by incfly** [comment]
Thanks for confirming. Just learned that the label filtering out the config control plane sees.

File: pilot/pkg/security/authn/v1beta1/policy_applier.go [hunk 1]
```diff
@@ -353,26 +352,33 @@ func getMutualTLSMode(mtls *v1beta1.PeerAuthentication_MutualTLS) model.MutualTL
...
-	workloadPolicyCreationTime := time.Unix(1<<63-1, 0)
+	var meshCfg, namespaceCfg, workloadCfg *model.Config
 
 	for _, cfg := range configs {
 		spec := cfg.Spec.(*v1beta1.PeerAuthentication)
-		if cfg.Namespace == rootNamespace && spec.Selector == nil {
-			meshPolicy = spec
-		} else if spec.Selector == nil {
-			namespacePolicy = spec
+		if spec.Selector == nil || len(spec.Selector.MatchLabels) == 0 {
```
Comment (incfly) [inline pilot/pkg/security/authn/v1beta1/policy_applier.go]: This config hierachy case switch has occurred several times in previous PR also. Having a utility func accept config and return the type (root/ns/wl) might be clearer. (optional)

Comment (rshriram) [reply pilot/pkg/security/authn/v1beta1/policy_applier.go]: I would suggest you really look at the sidecar API that resolves all of this way before the XDS request processing. Not really a fan of timestamp.before and after. A simple sort and picking the first non label policy, followed by breaking out of the for loop is easier. IMO, the more this logic diverges from how the sidecar API is done in push context, the more bugs you will hit as you hit corner cases in future.

**Pull request merged**